contracts/Redemption.vy

```diff
@@ -0,0 +1,495 @@
+# in 20 decimals for n >= 2
+X0: constant(int256)  = 128 * E18 # 18 decimals
+A0: constant(int256)  = 38_877_084_059_945_950_922_200 * E15 * E18 # no decimals
+X1: constant(int256)  = X0 / 2 # 18 decimals
+A1: constant(int256)  = 6_235_149_080_811_616_882_910 * E6 # no decimals
+X2: constant(int256)  = X1 * 100 / 2
+A2: constant(int256)  = 7_896_296_018_268_069_516_100 * E12
+X3: constant(int256)  = X2 / 2
+A3: constant(int256)  = 888_611_052_050_787_263_676 * E6
+X4: constant(int256)  = X3 / 2
+A4: constant(int256)  = 298_095_798_704_172_827_474 * E3
+X5: constant(int256)  = X4 / 2
+A5: constant(int256)  = 5_459_815_003_314_423_907_810
+X6: constant(int256)  = X5 / 2
+A6: constant(int256)  = 738_905_609_893_065_022_723
+X7: constant(int256)  = X6 / 2
+A7: constant(int256)  = 271_828_182_845_904_523_536
+X8: constant(int256)  = X7 / 2
+A8: constant(int256)  = 164_872_127_070_012_814_685
+X9: constant(int256)  = X8 / 2
+A9: constant(int256)  = 128_402_541_668_774_148_407
+X10: constant(int256) = X9 / 2
+A10: constant(int256) = 11_331_4845_306_682_631_683
+X11: constant(int256) = X10 / 2
+A11: constant(int256) = 1_064_49_445_891_785_942_956
+
+@external
+def __init__(
+    yfi: address, d_yfi: address, ve_yfi: address, owner: address, 
+    price_feed: address, curve_pool: address, scaling_factor: uint256,
+):
+    YFI = ERC20(yfi)
+    DYFI = IDYFI(d_yfi)
+    VEYFI = ERC20(ve_yfi)
+    PRICE_FEED = AggregatorV3Interface(price_feed)
+    CURVE_POOL = CurvePoolInterface(curve_pool)
+    self._transfer_ownership(owner)
+    self.payee = owner
+    self.packed_scaling_factor = shift(scaling_factor, 128) | shift(scaling_factor, 192)
+
+
+@payable
+@external
+def redeem(amount: uint256, recipient: address = msg.sender) -> uint256:
+    """
+    @notice Redeem your dYFI for YFI using ETH.
+    @dev Redemption tolerates a 0.3% negative or positive slippage.
+    @param amount amount of dYFI to spend
+    @param recipient of the exercised YFI
+    """
+    self._check_killed()
+    assert YFI.balanceOf(self) >= amount, "not enough YFI"
+    eth_required: uint256 = self._eth_required(amount)
+    assert eth_required > 0
+    tolerance: uint256 = eth_required * SLIPPAGE_TOLERANCE / SLIPPAGE_DENOMINATOR
+    if msg.value < (eth_required - tolerance) or msg.value > (eth_required + tolerance):
+        raise "price out of tolerance"
+    DYFI.burn(msg.sender, amount)
+    raw_call(self.payee, b"", value=msg.value)
+    YFI.transfer(recipient, amount)
+    return amount
+
+
+@external
+@view
+def discount() -> uint256:
+    """
+    @notice Get the current dYFI redemption discount
+    @return Redemption discount (18 decimals)
+    """
+    return self._discount()
+
+
+@internal
+@view
+def _discount() -> uint256:
+    yfi_supply: uint256 = YFI.totalSupply()
+    veyfi_supply: uint256 = VEYFI.totalSupply()
+    x: int256 = convert(veyfi_supply * UNIT / yfi_supply, int256)
+    x = self._exp(47 * (self._scaling_factor()[0] * x / E18 - E18) / 10)
```

Comment:
What is the expectation of discount calculation when  `self._scaling_factor()[0]` is zero value?

when the contract hasnt been thru ramp process can you redeem? im guessing no, since the contract needs YFI setup and DYFI,

but wondering if theres ever a state that would make self._scaling_factor()[0] = 0 and if the calculation discount is expected to be non zero, since doing quick math by replacing factors i think is a value above 0 which may be dust, but wanted to be sure if expected

Reply:
`self._scaling_factor()[0]` is never zero, see `start_ramp()`. it requires the value is between 1 and 12. It will be deployed with a value of 10. I will add the same requirement in the constructor to guarantee it is never zero.